clients/src/main/java/org/apache/kafka/clients/consumer/CooperativeStickyAssignor.java

```diff
@@ -0,0 +1,114 @@
+ * needlessly had its partitions revoked and reassigned. Incremental cooperative rebalancing tries to improve the total
+ * rebalancing latency by allowing a consumer to hang on to its owned partitions before sending the JoinGroup, and only
+ * revoke those partitions that will ultimately be assigned to another consumer after the rebalance.
+ *
+ *
+ * To turn on cooperative rebalancing you must set all your consumers to use this {@code PartitionAssignor},
+ * or implement a custom one that returns {@code RebalanceProtocol.COOPERATIVE} in
+ * {@link CooperativeStickyAssignor#supportedProtocols supportedProtocols()}.
+ * <p>
+ * IMPORTANT: if upgrading from 2.3 or earlier, you must follow a specific upgrade path in order to safely turn on
+ * cooperative rebalancing. See the upgrade guide for details.
+ */
+public class CooperativeStickyAssignor extends AbstractStickyAssignor {
+
+    @Override
+    public String name() {
+        return "cooperative";
+    }
+
+    @Override
+    protected MemberData memberData(Subscription subscription) {
+        return new MemberData(subscription.ownedPartitions(), Optional.empty());
+    }
+
+    @Override
+    public Map<String, List<TopicPartition>> assign(Map<String, Integer> partitionsPerTopic,
+                                                    Map<String, Subscription> subscriptions) {
+
+        final Map<String, List<TopicPartition>> assignments = super.assign(partitionsPerTopic, subscriptions);
+        adjustAssignment(subscriptions, assignments);
+        return assignments;
+    }
+
+    // Following the cooperative rebalancing protocol requires removing partitions that must first be revoked from the assignment
+    private void adjustAssignment(final Map<String, Subscription> subscriptions,
+                                  final Map<String, List<TopicPartition>> assignments) {
+
+        Map<TopicPartition, String> allOwnedPartitions = new HashMap<>();
+        Map<TopicPartition, String> allAssignedPartitions = new HashMap<>();
+        Set<TopicPartition> allRevokedPartitions = new HashSet<>();
+
+        for (final Map.Entry<String, List<TopicPartition>> entry : assignments.entrySet()) {
+            final String consumer = entry.getKey();
+
+            final List<TopicPartition> ownedPartitions = subscriptions.get(consumer).ownedPartitions();
+            for (TopicPartition tp : ownedPartitions) {
+                allOwnedPartitions.put(tp, consumer);
+            }
+
+            final List<TopicPartition> assignedPartitions = entry.getValue();
+            for (TopicPartition tp : assignedPartitions) {
+                allAssignedPartitions.put(tp, consumer);
+            }
+
+            final Set<TopicPartition> revokedPartitions = new HashSet<>(ownedPartitions);
+            revokedPartitions.removeAll(assignedPartitions);
+            allRevokedPartitions.addAll(revokedPartitions);
+        }
+
+        // for all owned but not assigned partitions, blindly add them to assignment
```

Comment:
Would this ever happen in `AbstractStickyAssignor` though (note we write the code so it's really under our full control :P)? I think it should not happen, so if it ever happens it would be a bug?

Reply:
I actually took this out, because it's not correct to do so (I had just copied it over from the old `adjustAssignment` implementation from Pt 2 before we took it out

Reply:
For example, if the subscription changes a partition may be in ownedPartitions but definitely should not be in assignment. What was the original reasoning for doing this?

Reply:
If the subscription changes, the `onPartitionsRevoked` would be triggered and the owned-partitions in subscription would not include them.

The original reasoning is just to make sure all partitions except the ones that are attempted for reassignment-hence-revocation-first are owned by someone. But like we discussed in the other PR, today we are not checking these conditions anyways, e.g.:

1. assignor assign one partition to multiple consumers.
2. assignor assign one partition to no one.

They are both bad situations but Coordinator does not check for those today. We discussed about this and think it worth a more general solution as a separate JIRA.